File: 68_q31_01.py
square = 6
count = 0
is_used = [[[False, False] for _ in range(square + 1)] for _ in range(square + 1)]

def route(x, y, is_first_time):
    global count
    if x == square and y == square:
        if is_first_time:
            route(0, 0, False)
        else:
            count += 1
    if x < square:
        if not is_used[x][y][0]:
            is_used[x][y][0] = True
            route(x + 1, y, is_first_time)
            is_used[x][y][0] = False
    if y < square:
        if not is_used[x][y][1]:
            is_used[x][y][1] = True
            route(x, y + 1, is_first_time)
            is_used[x][y][1] = False

route(0, 0, True)
print(count)

File: 68_q31_01.cpp
#include <iostream>
#include <vector>

int square = 6;
int count = 0;
std::vector<std::vector<std::vector<bool>>> is_used(square + 1, std::vector<std::vector<bool>>(square + 1, std::vector<bool>(2, false)));

void route(int x, int y, bool is_first_time) {
    if (x == square && y == square) {
        if (is_first_time) {
            route(0, 0, false);
        } else {
            count++;
        }
    }
    if (x < square) {
        if (!is_used[x][y][0]) {
            is_used[x][y][0] = true;
            route(x + 1, y, is_first_time);
            is_used[x][y][0] = false;
        }
    }
    if (y < square) {
        if (!is_used[x][y][1]) {
            is_used[x][y][1] = true;
            route(x, y + 1, is_first_time);
            is_used[x][y][1] = false;
        }
    }
}

int main() {
    route(0, 0, true);
    std::cout << count << std::endl;
    return 0;
}